scarlet/transformations.py

```diff
@@ -15,147 +16,437 @@ def check_cache(name, key):
-    operators translating the PSF convolution kernel, making a single transformation
-    Gamma = Ty.P.Tx, where Tx,Ty are the translation operators and P is the PSF
-    convolution operator.
+    A `LinearFilter` is defined by an image of weights
+    and a list of relative coordinates from the current pixels
+    for each weight.
+    This method converts those coordinates into slices that are
+    used to apply the filter.
     """
-    def __init__(self, psf=None):
-        """Constructor
+    slices = []
+    inv_slices = []
+    for cy, cx in coords:
+        _slice = [slice(None),slice(None)]
+        _inv_slice = [slice(None),slice(None)]
+        if cy>0:
+            _slice[0] = slice(cy,None)
+            _inv_slice[0] = slice(None,-cy)
+        elif cy<0:
+            _slice[0] = slice(None,cy)
+            _inv_slice[0] = slice(-cy, None)
+        if cx>0:
+            _slice[1] = slice(cx,None)
+            _inv_slice[1] = slice(None,-cx)
+        elif cx<0:
+            _slice[1] = slice(None,cx)
+            _inv_slice[1] = slice(-cx, None)
+        slices.append(_slice)
+        inv_slices.append(_inv_slice)
+    return slices, inv_slices
+
+def apply_filter(X, weights, slices, inv_slices):
+    """Apply a filter to a 2D image X
+
+    Parameters
+    ----------
+    X: 2D numpy array
+        The image to apply the filter to
+    weights: 1D array
+        Weights corresponding to each slice in `slices`
+    slices: list of `slice` objects
+        Slices in the new `X` to store the filtered X
+    inv_slices: list of `slice` objects
+        Slices of `X` to apply each weight
+    
+    Returns
+    -------
+    new_X: 2D numpy array
+        The result of applying the filter to `X`
+    """
+    result = np.zeros(X.shape, dtype=X.dtype)
+    for n, weight in enumerate(weights):
+        result[slices[n]] += weight * X[inv_slices[n]]
+    return result
+
+class LinearFilter:
+    """A filter that can be applied to an image
+
+    This acts like a sparse diagonal matrix that applies an
+    image of weights to a 2D matrix.
+    """
+    def __init__(self, img, coords):
+        """Initialize the Filter
 
         Parameters
         ----------
-        psf: array-like, default=`None`
-            PSF image in either a single band (used for all images)
-            or an array/list of images with a PSF image for each band.
+        img: 2D or 1D array-like
+            Weights to apply to the filter
+        coords: 2D or 1D array-like
+            Relative coordinates from the current pixel
+            for each weight in `img`
+            (so [0,0] is the current pixel).
         """
-        self.psf = psf
-        self.cache = {}
-
-    def _make_matrices(self, shape, offset_int):
-        """Build Tx, Ty, P, Gamma
+        self.img = np.array(img)
+        self._coords = np.array(coords)
+        self.slices, self.inv_slices = get_filter_slices(self._coords.reshape(-1, 2))
```

Comment:
This will not work if `img` is in fact an image, in which case `coords` are redundant. This might also get potentially inconsistent with `self.img.reshape` in L136.

Reply:
Not quite. There are two options here: 
1) `img` is a 2D filter image, then it need no coordinates, but constant reshaping in `dot()`
2) `img` is a 1D list of values, together with a list of 2D coordinates, in which the coordinates are necessary, but there is no need for reshaping in `dot()`.

I think we should store internally the latter version and make sure that we can parse both.